packages/web/src/sideEffects/localFs/utils.js

```diff
@@ -1,18 +1,50 @@
+/*
+ * Compare the given files for equality; file path and file contents
+ * @return {Boolean} true if the files are the same
+ */
+const sameFile = (a, b) => {
+  if (a.fullPath === b.fullPath) {
+    if (typeof a.source === typeof b.source) {
+      if (a.source.byteLength) {
+        // compare ArrayBuffer contents
+        const aArray = new Uint8Array(a.source)
+        const bArray = new Uint8Array(b.source)
+        return aArray.length === bArray.length && aArray.every((v, i) => v === bArray[i])
+      } else {
+        // compare String contents
+        return a.length === b.length && a.source === b.source
+      }
+    }
+  }
+  return false
+}
+
 /**
- * returns list of files whose source has changed compared to the file with the same
- * name in memFs
- * @param {Object} reference - The reference hash of files, organized by name.
- * @param {Array} files - The list of files : need to have a source & name field.
- * @return array of files that changed
+ * Compare the contents of the reference and the new file lists.
+ * @see flattenFiles() below
+ * @param {Array} reference - the reference list of files
+ * @param {Array} files - the new list of files
+ * @return {Array} a list of the changed files
  */
-const changedFiles = (reference, files) => files.filter((file) => {
-  const matches = reference.filter((ref) => ref.fullPath === file.fullPath && ref.source !== file.source)
-  return matches.length > 0
-})
+const changedFiles = (reference, files) =>  {
+  const working = reference.slice()
+  return files.filter((file) => {
+    const found = working.findIndex((a, i) => sameFile(a, file))
```

Comment:
also feel like this could be quite inefficient. It's doing full file comparisons over and over.  Wonder if we can hash and cache, or something similar.

Reply:
The list of references is reduced if there’s a match. Not perfect but far better then before. It actually works too. 👍